### ISSUE TITLE
In order to [enable grouped orders] As a [client and restaurant and courier] I want [a new feature for clients to create a groupe and then order with this group]

### ISSUE THREAD

**Issue by Doyoubouilloux** [coopcycle/coopcycle-web#1456]
**In order to**  `enable grouped orders`
Enable clients working in companies to create a group on the platform resto.paris and then order with this group. No one has to pay for the group, each person within the group pay for its own order witch is then added on the group order.

**As a** `client`
The client won't have to pay for the whole group and can benefit from the grouped order to have free delivery (because the group order will be above a certain amount of money)

**As a** `restaurant and courier`
The restaurant and courier will have regular orders from companies and will be able to organize its work better in advanced. It is secure income as well.

**I want** `a feature enablelling grouped orders`
One client must be able to create a group and invite its colleagues
Each member of the group must be able to créate a group order and then pay for their own order
The order must be validated once the order reaches a certain amount of money. Then the delivery is organised.
Each group = one adress = a perimeter with restaurants where they can order

**Whereas currently** `one client has to take responsability for the group order`
Now, someone in the company has to pay the order for the whole groupe with its personal account...

**Additional context**
Feature required for resto.paris in order for the plateform to make a difference and reach companies need
[Resto.paris_spec commande groupée.pdf](https://github.com/coopcycle/coopcycle-web/files/4718093/Resto.paris_spec.commande.groupee.pdf)

**Comment by alexsegura** [comment]
Basically, a customer may start an order, and invite friends to add items to this same order. 

### Implementing the shareable link

An order can be shared via a JSON Web Token. The JWT contains custom claims with: 
- the order id
- the email of the guest
- (maybe also the id of the "host" customer)

Use the example below on [jwt.io](https://jwt.io/)

```
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJvcmRlciI6MSwiZ3Vlc3QiOiJtZXguemt0a0BnbWFpbC5jb20ifQ.uMwUEU__oT4MsL1B4q6y-iHko45E3APQS92CW6EN5IE
```

This allows us to restore the order from database easily. 
The shareable link will be like `/order/share?token=${jwt}`

The shareable link does not require to have an account on the website, or to be signed in, anybody having the link can use it. 

### Changes to the order model

At least, we need to add a `BOOLEAN shared` column to the order, to mark it as shared. 
Also, we need to add a new entity / table `OrderGuest` / `sylius_order_guest (order_id, email)` (oneToMany), that just keeps track of the guests. 

The `shared` property switches to true when the email has been sent to guests. 

Basically, the UX flow is like: 

- A customer goes to a restaurant page
- He/she adds AT LEAST ONE item to the cart
- A button "Invite friends to add items" appears
- He/she can add one or multiple emails, and click "Invite"
- An email is sent to each guest

### Restoring the order from the JWT

When the shareable link is invoked:

- Decode the JWT (error 400 if the signature is invalid, or JWT expired)
- Check if it contains the expected claims (error 400 if not)
- Check if shared = true 
- Retrieve the order from database
- Check if is in state = cart (error 400 if not)
- If all OK, set the cart in session (like [here](https://github.com/coopcycle/coopcycle-web/blob/88365a37baf937d65d614c41ef15712524a036b7/src/Controller/OrderController.php#L346-L385)), and redirect to the restaurant page

---

### Limitations for guests

#### Guests can't change the order address & customer

When the order is shared = true AND the current user is NOT the host, the address field should be disabled. 

#### Guests can't continue the checkout

Guests can't click the "Order" button, to begin the checkout process. Instead, when they click this button, it means "I have finished adding my items". 

In `OrderController`, some access control should be implemented to prevent this (in case they enter the `/order/` URL manually)

---

### Notifications between host & guests

When guests add items to the order, the host should see the items updating "live". 
When guests click on the "Order" button, the host should see a small notification indicating this. 

--- 

⚠️ On the "MVP" implementation, guests can modify all the items of the order, and we don't keep track of who added what. 

A more advanced implementation, to keep track of who added what, could be implemented via an "extension" to the `OrderItem` entity. 

Basically, we add a `$owner` property to the `OrderItem` entity (a table like `(order_item_id, email, created_at)`. 
When a guest adds an item to the cart, we do `$orderItem->setOwner($email);`.
By default, the `$owner` is NULL, indicating the item "belongs" to the host. 

This would allow to group items per user/guest, and to forbid people to remove other's items.

**Comment by alexsegura** [comment]
@moeffju Hi! Sorry for the veeeery late response! Almost 3 years later, we are just starting implementing the feature. If you want, you can fund us on [OpenCollective](https://opencollective.com/coopcycle)

**Comment by alexsegura** [comment]
We will actually need to keep track of which items belongs to who, because we want to show who owes who. 

Currently, when adding the same product/variant to the cart, it increments the quantity. The implementation should allow adding the same product/variant **twice**, to easily distinguish between customers (see example below). 

<img width="371" alt="Capture d’écran 2023-03-22 à 09 15 47" src="https://user-images.githubusercontent.com/1162230/226840717-5144fe00-094c-4cef-b875-bc5a48eaaa7b.png">

Basically,  the `OrderItem::equals` methods should also check who "owns" the item. It should check if the 

https://github.com/coopcycle/coopcycle-web/blob/56df3cdf7fcba5b657fb54e34aecc4cc773174f1/src/Entity/Sylius/OrderItem.php#L51-L54
